daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala

```diff
@@ -24,46 +27,113 @@ private[digitalasset] object DamlLfEncoder extends App {
 
   private val pkgId = Ref.PackageId.assertFromString("-self-")
 
-  try {
-    val (inputFile, outputFile, languageVersion): (String, String, LanguageVersion) =
-      args.toSeq match {
-        case Seq(input, output) =>
-          (input, output, LanguageVersion.default)
-        case Seq(input, output, version) =>
-          version.split("""\.""").toSeq match {
-            case Seq("1", minor)
-                if LanguageMajorVersion.V1.supportsMinorVersion(minor) || minor == "dev" =>
-              (input, output, LanguageVersion(LanguageMajorVersion.V1, minor))
-            case _ =>
-              error(s"version '$version' not supported")
-          }
-        case _ =>
-          error("usage: damlf-as input output [version]")
-      }
-
-    implicit val parserParameters: ParserParameters[this.type] =
-      ParserParameters(
-        defaultPackageId = pkgId,
-        languageVersion = languageVersion
-      )
-
-    val string = Source.fromFile(Paths.get(inputFile).toFile, "UTF8").mkString
-
-    val modules = parseModules[this.type](string).fold(error, identity)
+  private def main() =
+    try {
+      val appArgs = parseArgs()
+
+      implicit val parserParameters: ParserParameters[this.type] =
+        ParserParameters(
+          defaultPackageId = pkgId,
+          languageVersion = appArgs.languageVersion
+        )
+
+      makeDar(readSources(appArgs.inputFiles), Paths.get(appArgs.outputFile).toFile)
+
+    } catch {
+      case e: EncodeError =>
+        error(s"Encoding error: ${e.message}")
+      case NonFatal(e) =>
+        error(s"error: ${e.getMessage}")
+    }
+
+  private def readSources(files: Seq[String]) =
+    files.flatMap(file => Source.fromFile(Paths.get(file).toFile, "UTF8")).mkString
```

Comment:
`flatMap` and `mkString` will make an intermediate `Seq[Char]` all boxed first 😨 
```suggestion
    files.map(file => Source.fromFile(Paths.get(file).toFile, "UTF8")).mkString
```

Reply:
sorry !